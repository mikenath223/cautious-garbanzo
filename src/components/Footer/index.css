.footer {
  text-align: center;
  display: flex;
  justify-content: space-between;
  padding: 2rem;
}
