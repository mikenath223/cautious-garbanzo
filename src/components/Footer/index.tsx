import "./index.css";

export default function Footer() {
  return (
    <footer className="p-5">
      <div className="footer max-width-inhibitor">
        <h5>WeatherAPP {`${new Date().getFullYear()}`} &copy;</h5>
        <span className="italic font-medium">Powered by weatherstack</span>
      </div>
    </footer>
  );
}
