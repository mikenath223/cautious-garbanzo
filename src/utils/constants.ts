import { IWeatherInfo } from "./types"

export const initFavCities: string[] = [
  "Tokyo",
  "Delhi",
  "Shanghai",
  "Dhaka",
  "Sao Paulo",
  "Mexico City",
  "Cairo",
  "Beijing",
  "Mumbai",
  "Osaka",
  "Chongqing",
  "Karachi",
  "Kinshasa",
  "Lagos",
  "Istanbul"
]

export const defaultCity: IWeatherInfo = {
  "request": {
    "type": "City",
    "query": "New York, United States of America",
  },
  "location": {
    "name": "New York",
    "country": "United States of America",
    "region": "New York",
    "localtime": "2019-09-07 08:14",
  },
  "current": {
    "observation_time": "12:14 PM",
    "temperature": 13,
    "weather_code": 113,
    "weather_icons": [
      "https://assets.weatherstack.com/images/wsymbols01_png_64/wsymbol_0001_sunny.png"
    ],
    "weather_descriptions": [
      "Sunny"
    ],
    "wind_speed": 0,
    "wind_degree": 349,
    "wind_dir": "N",
    "pressure": 1010,
    "precip": 0,
    "humidity": 90,
    "cloudcover": 0,
    "feelslike": 13,
    "uv_index": 4,
    "visibility": 16
  }
}

export const defaultNote = {
  id: "01",
  text: "Welcome"
}
