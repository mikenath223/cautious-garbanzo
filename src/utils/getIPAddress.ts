const fetchIPAddress = async () => {
  try {
    const response = await fetch('https://ipapi.co/json')
    const data = await response.json()
    return data
  } catch (err) {
    console.error(err)
  }
}

export default fetchIPAddress
