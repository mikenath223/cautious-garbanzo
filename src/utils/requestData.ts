import { IWeatherInfo, PromiseFulfilledResult } from "./types"

export const getCityWeather = async (city: string | null = null, isCheckCache: boolean, lat: string | null = null, lon: string | null = null): Promise<IWeatherInfo> => {
  try {
    const currentDate = new Date()
    const localStorageKey = `${city?.toLowerCase()?.trim() || (lat || "" + lon || "")}
    -${currentDate.getHours()}-${currentDate.getDate()}-${currentDate.getMonth()}
    -${currentDate.getFullYear()}`
    const cachedCityData = localStorage.getItem(localStorageKey)
    if (cachedCityData && isCheckCache) {
      const weatherData = JSON.parse(cachedCityData)
      return weatherData
    }
    const url = getQueryUrl(city, lat, lon)
    const response = await fetch(url)
    const result = await response.json()
    localStorage.setItem(localStorageKey, JSON.stringify(result))
    return result
  } catch (error) {
    console.log(error)
    const errMsg = `Failed to retrieve ${city?.trim()} data!`
    throw new Error(errMsg)
  }
}

const sortCities = (cities: string[]) => cities.sort((a, b) => a.toLowerCase().localeCompare(b.toLowerCase()))

export const getFavsCityWeather = async (cities: string[]) => {
  try {
    const fetchRequests = sortCities(cities).map(city => getCityWeather(city, true));
    const citiesInfo = await Promise.allSettled(fetchRequests)
    const data = citiesInfo.filter(data => data.status === "fulfilled").map(data => (data as PromiseFulfilledResult<IWeatherInfo>).value)
    return data
  } catch (_) {
    return []
  }
}

const getQueryUrl = (city: string | null, lat: string | null, lon: string | null) => {
  const cityQuery = `&q=${city}`
  const latQuery = `&lat=${lat}&lon=${lon}`
  if (city && lat) {
    return `https://api.openweathermap.org/data/2.5/weather?units=metric${cityQuery}${latQuery}&APPID=${process.env.REACT_APP_OPENWEATHER_API_KEY}`
  } else if (!lat || !lon) {
    return `https://api.openweathermap.org/data/2.5/weather?units=metric${cityQuery}&APPID=${process.env.REACT_APP_OPENWEATHER_API_KEY}`
  }
  return `https://api.openweathermap.org/data/2.5/weather?units=metric${latQuery}&APPID=${process.env.REACT_APP_OPENWEATHER_API_KEY}`
}
