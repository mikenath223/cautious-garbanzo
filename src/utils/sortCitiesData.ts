import { IWeatherInfo } from "./types";

const sortCitiesData = (cities: IWeatherInfo[] | undefined) => cities?.sort((a, b) => a.location.name.toLowerCase().localeCompare(b.location.name.toLowerCase()))

export default sortCitiesData
