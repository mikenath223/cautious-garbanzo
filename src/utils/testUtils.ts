import { render } from '@testing-library/react';

export const withRender = (component: JSX.Element) =>
  render(component);

export const runMock = () => {
  const mockSuccessResponse = {};
  const mockJsonPromise = Promise.resolve(mockSuccessResponse);
  const mockFetchPromise = Promise.resolve({
    ok: true,
    json: () => mockJsonPromise,
  });

  global.fetch = jest.fn(() => mockFetchPromise) as jest.Mock;
  jest.spyOn(localStorage, "setItem");
  localStorage.setItem = (key: string, value: string) => { };
  localStorage.getItem = (key: string) => null;
}