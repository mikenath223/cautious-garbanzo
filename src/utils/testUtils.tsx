import { render } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";

export const withRender = (component: JSX.Element) =>
  render(<MemoryRouter>{component}</MemoryRouter>);

export const runMock = () => {
  const mockSuccessResponse = {};
  const mockJsonPromise = Promise.resolve(mockSuccessResponse);
  const mockFetchPromise = Promise.resolve({
    ok: true,
    json: () => mockJsonPromise,
  });

  global.fetch = jest.fn(() => mockFetchPromise) as jest.Mock;
  jest.spyOn(localStorage, "setItem");
  localStorage.setItem = (key: string, value: string) => {};
  localStorage.getItem = (key: string) => null;
};
