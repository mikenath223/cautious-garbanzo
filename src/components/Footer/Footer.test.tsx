import { withRender } from "../../utils/testUtils";
import Footer from "./index";

describe("Footer", function () {
  it("should match snapshot", () => {
    const { container } = withRender(<Footer />);

    expect(container).toMatchSnapshot();
  });
});
