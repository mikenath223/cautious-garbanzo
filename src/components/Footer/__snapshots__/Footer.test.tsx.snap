// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Footer should match snapshot 1`] = `
<div>
  <footer
    class="p-5"
  >
    <div
      class="footer max-width-inhibitor"
    >
      <h5>
        WeatherAPP 
        2023
         Â©
      </h5>
      <span
        class="italic font-medium"
      >
        Powered by OpenWeather
      </span>
    </div>
  </footer>
</div>
`;
